File: overlay/python/hydra-kernel/src/main/python/hydra/overlay/python/lib/regex.py
"""Python implementations of hydra.lib.regex primitives.

Patterns are Hydra-defined and translingual (docs/specification/regex.md). Each primitive first runs
the pattern through hydra.parse.regex, then renders the AST to PCRE syntax via hydra.print.pcre.regex
(Python's re module is PCRE-like), before handing the rendered pattern to the native engine. An
ill-formed pattern (rejected by hydra.parse.regex) is treated as "no match" -- the same portable-
failure convention as an empty match. See issue #603.
"""

from __future__ import annotations

import re

import hydra.parse.regex
import hydra.print.pcre.regex
from hydra.overlay.python.dsl.python import Given, None_, Optional, NONE_, frozenlist


def _to_native(pattern: str) -> Optional[str]:
    """Translate a Hydra regex pattern to PCRE syntax. None_() if the pattern does not parse."""
    match hydra.parse.regex.parse_regex(pattern):
        case Given(value=ast):
            return Given(hydra.print.pcre.regex.print_regex(ast))
        case None_():
            return NONE_


def matches(pattern: str, input: str) -> bool:
    """Check whether an entire string matches a regex pattern."""
    match _to_native(pattern):
        case Given(value=native):
            return re.fullmatch(native, input) is not None
        case None_():
            return False


def find(pattern: str, input: str) -> Optional[str]:
    """Find the first substring matching a regex pattern."""
    match _to_native(pattern):
        case Given(value=native):
            m = re.search(native, input)
            if m is not None:
                return Given(m.group())
            return NONE_
        case None_():
            return NONE_


def find_all(pattern: str, input: str) -> frozenlist[str]:
    """Find all non-overlapping substrings matching a regex pattern."""
    match _to_native(pattern):
        case Given(value=native):
            return tuple(re.findall(native, input))
        case None_():
            return ()


def replace(pattern: str, replacement: str, input: str) -> str:
    """Replace the first occurrence of a regex pattern with a replacement string."""
    match _to_native(pattern):
        case Given(value=native):
            return re.sub(native, replacement, input, count=1)
        case None_():
            return input


def replace_all(pattern: str, replacement: str, input: str) -> str:
    """Replace all non-overlapping occurrences of a regex pattern with a replacement string."""
    match _to_native(pattern):
        case Given(value=native):
            return re.sub(native, replacement, input)
        case None_():
            return input


def split(pattern: str, input: str) -> frozenlist[str]:
    """Split a string by a regex pattern."""
    match _to_native(pattern):
        case Given(value=native):
            return tuple(re.split(native, input))
        case None_():
            return (input,)

File: overlay/haskell/hydra-kernel/src/main/haskell/Hydra/Overlay/Haskell/Lib/Regex.hs
-- | Haskell implementations of hydra.lib.regex primitives.
--
-- Patterns are Hydra-defined and translingual (docs/specification/regex.md): each primitive first
-- runs the pattern through hydra.parse.regex, then renders the resulting AST to POSIX ERE syntax via
-- hydra.print.posix.regex (Haskell's Text.Regex.TDFA is POSIX ERE), and only then hands the rendered
-- pattern to the native engine. An ill-formed pattern (one that hydra.parse.regex rejects) is treated
-- as "no match" — the same portable-failure convention as an empty match, so no primitive here raises
-- an error for a malformed pattern; see issue #603.

module Hydra.Overlay.Haskell.Lib.Regex where

import qualified Text.Regex.TDFA as R

import qualified Hydra.Parse.Regex as ParseRegex
import qualified Hydra.Print.Posix.Regex as PrintPosixRegex


-- | Translate a Hydra regex pattern to POSIX ERE syntax via parse.regex |> print.posix.regex. Nothing
-- if the pattern does not parse (ill-formed).
toNative :: String -> Maybe String
toNative pattern = PrintPosixRegex.printRegex <$> ParseRegex.parseRegex pattern

-- | Find the first substring matching a regex pattern.
find :: String -> String -> Maybe String
find pattern input = case toNative pattern of
  Nothing -> Nothing
  Just native -> case (input R.=~ native :: (String, String, String)) of
    (_, match, _)
      | null match -> Nothing
      | otherwise  -> Just match

-- | Find all non-overlapping substrings matching a regex pattern.
findAll :: String -> String -> [String]
findAll pattern input = case toNative pattern of
  Nothing -> []
  Just native -> R.getAllTextMatches (input R.=~ native :: R.AllTextMatches [] String)

-- | Check whether an entire string matches a regex pattern.
matches :: String -> String -> Bool
matches pattern input = case toNative pattern of
  Nothing -> False
  Just native -> input R.=~ ("^(" ++ native ++ ")$" :: String) :: Bool

-- | Replace the first occurrence of a regex pattern with a replacement string.
replace :: String -> String -> String -> String
replace pattern replacement input = case toNative pattern of
  Nothing -> input
  Just native -> case (input R.=~ native :: (String, String, String)) of
    (before, match, after)
      | null match -> input
      | otherwise  -> before ++ replacement ++ after

-- | Replace all non-overlapping occurrences of a regex pattern with a replacement string.
replaceAll :: String -> String -> String -> String
replaceAll pattern replacement input = case toNative pattern of
  Nothing -> input
  Just native -> go input
    where
      go "" = ""
      go s = case (s R.=~ native :: (String, String, String)) of
        (before, match, after)
          | null match -> s
          | otherwise  -> before ++ replacement ++ go after

-- | Split a string by a regex pattern.
split :: String -> String -> [String]
split pattern input = case toNative pattern of
  Nothing -> [input]
  Just native -> go input
    where
      go "" = [""]
      go s = case (s R.=~ native :: (String, String, String)) of
        (before, match, after)
          | null match -> [s]
          | otherwise  -> before : go after
